atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -728,19 +736,34 @@ private Comparator<Cell> preserveInputRowOrder(List<Map.Entry<Cell, Value>> inpu
      * per non-empty row.
      */
     private Iterator<Map.Entry<byte[], RowColumnRangeIterator>> partitionByRow(
-            Iterator<Map.Entry<Cell, Value>> rawResults) {
+            Iterator<Map.Entry<Cell, Value>> rawResults, Iterator<byte[]> expectedRows) {
         PeekingIterator<Map.Entry<Cell, Value>> peekableRawResults = Iterators.peekingIterator(rawResults);
-        return new AbstractIterator<Map.Entry<byte[], RowColumnRangeIterator>>() {
+        return new AbstractIterator<>() {
             byte[] prevRowName;
 
             @Override
             protected Map.Entry<byte[], RowColumnRangeIterator> computeNext() {
                 finishConsumingPreviousRow(peekableRawResults);
-                if (!peekableRawResults.hasNext()) {
+                if (!expectedRows.hasNext()) {
+                    if (peekableRawResults.hasNext()) {
+                        throw new SafeIllegalStateException(
```

Comment:
Any reason why not to use `Preconditions.checkState` here?

Reply:
`Preconditions.checkState` eagerly evaluates its arguments, so we would get a `NoSuchElementException` on the happy path when we try to peek as part of building the `UnsafeArg`.